### PR TITLE
LPD-3281 - It's possible to insert via API negative numeric values on the product SKU's Details tab

### PR DESCRIPTION
**Pull request by matyaswollner**
https://liferay.atlassian.net/browse/LPD-3281

### CODE REVIEW

**Comment by matyaswollner** [comment]
ci:test:sf

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 1]
```diff
@@ -2058,6 +2070,40 @@ private void _validateSku(
 			"Duplicate SKU value for product definition ID " + cpDefinitionId);
 	}
 
+	private void _validateSkuNumericValues(
+			double width, double height, double depth, double weight,
+			BigDecimal price, BigDecimal promoPrice, BigDecimal cost)
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:2075]: please sort the parameters alphabetically

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 2]
```diff
@@ -2058,6 +2070,40 @@ private void _validateSku(
...
+			throw new CPInstancePriceException();
+		}
+
+		if ((promoPrice != null) && (promoPrice.signum() < 0)) {
+			throw new CPInstancePromoPriceException();
+		}
+
+		if ((cost != null) && (cost.signum() < 0)) {
+			throw new CPInstanceCostException();
+		}
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:2104]: please sort the cases alphabetically

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 3]
```diff
@@ -2058,6 +2070,40 @@ private void _validateSku(
...
+
+		if (depth < 0) {
+			throw new CPInstanceDepthException();
+		}
+
+		if (weight < 0) {
+			throw new CPInstanceWeightException();
+		}
+
+		if ((price != null) && (price.signum() < 0)) {
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:2094]: to compare the BigDecimal use BigDecimalUtil.lt(price, BigDecimal.ZERO)

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 4]
```diff
@@ -2058,6 +2070,40 @@ private void _validateSku(
...
+			throw new CPInstancePriceException();
+		}
+
+		if ((promoPrice != null) && (promoPrice.signum() < 0)) {
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:2098]: same

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 5]
```diff
@@ -2058,6 +2070,40 @@ private void _validateSku(
...
+			throw new CPInstancePromoPriceException();
+		}
+
+		if ((cost != null) && (cost.signum() < 0)) {
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:2102]: same

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 6]
```diff
@@ -138,6 +145,8 @@ public CPInstance addCPInstance(
 		_validateExternalReferenceCode(
 			0, serviceContext.getCompanyId(), externalReferenceCode);
 		_validateSku(cpDefinitionId, 0, sku);
+		_validateSkuNumericValues(
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:148]: let's call this only _validate and move it as first

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 7]
```diff
@@ -1046,6 +1055,9 @@ public CPInstance updateCPInstance(
 
 		_validateSku(cpInstance.getCPDefinitionId(), cpInstanceId, sku);
 
+		_validateSkuNumericValues(
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:1058]: let's call this only _validate and move this as first (and remove empty lines between the _validate* statements)

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 8]
```diff
@@ -2058,6 +2070,40 @@ private void _validateSku(
 			"Duplicate SKU value for product definition ID " + cpDefinitionId);
 	}
 
+	private void _validateSkuNumericValues(
```
Comment (smottal) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:2073]: only _validate

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductResourceTest.java [hunk 9]
```diff
@@ -266,9 +326,118 @@ protected Product testPostProductClone_addProduct(Product product)
...
+			productResource.patchProduct(
+				postProduct.getProductId(), randomProduct);
+
+			Assert.fail();
+		}
+		catch (Problem.ProblemException problemException) {
+			Problem problem = problemException.getProblem();
+
+			Assert.assertEquals("BAD_REQUEST", problem.getStatus());
+
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductResourceTest.java:426]: remove this empty line

**Review by smottal — CHANGES_REQUESTED** [review]

**Review by alessiorendina — APPROVED** [review]
@matyaswollner @smottal Exception mappers look good to me!

**Comment by matyaswollner** [comment]
ci:test:sf

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:forward

**Pull request closed without merging**